feat: Make `apply` not evaluate arguments

laxsjo opened the pull request:
Transitioned `apply` to be a built-in and change its behavior to not evaluate it's arguments. Moves its documentation from dynref.lisp to lbmref.lisp, and update it to explain the new behavior.

Unfortunately this essentially required duplicating all branches in the `APPLICATION_START` continuation, implementing each kind of applicable value. So the binary size increased relatively much. When I compile the stormcore_100s target in the bldc repository with the changes from this commit the size went from 449092 to _449836_ bytes, an increase of 744. :/

I will try to optimize it further (hence the draft state). Please let me now if you spot any potential optimizations @svenssonjoel (for reference, you would want to be looking in 6e9f11c feat: Make `apply` a built-in)!

svenssonjoel commented:
Thanks @laxsjo,  

Cannot apply be defined something like this if you do it as a special form?   I am a bit surprised about the complexity in the apply_apply.    

``` 
/* special form apply */
 static void eval_apply(eval_context_t *ctx) {
   // some check on curr_exp; 
   lbm_value fun = lbm_car(ctx->curr_exp);
   lbm_value ls = lbm_car(lbm_cdr(ctx->curr_exp));

   lbm_value ap = lbm_cons(fun, ls); // check error

   ctx->curr_exp = ap;
 }
 ``` 

Thank you and have a nice day

svenssonjoel commented:
Thanks for the noreturn tip!

laxsjo commented on src/eval_cps.c:
```diff
@@ -2435,6 +2447,143 @@ static void cont_wait(eval_context_t *ctx) {
+ * @param curr_env The environment to re-evaluate the result of the macro
+ *   experssion in. 
+ */
+static void setup_macro(eval_context_t *ctx, lbm_value macro, lbm_value args, lbm_value curr_env) {
```
I have copied this comment thread from the commit 6e9f11c into this PR for easier consumption.

svenssonjoel wrote:
> I hope the compiler inlines these. If it actually generates a 4 argument function that could be a bit bad performance-wise. We can check the binary and if not inlined add the inline qualifier and look at the binary again.

laxsjo replied:
Oh, I refactored it out to this function specifically because I wanted it to not inline soo that we don't waste space.
But we can maybe get rid of the macro argument, and make it come from ctx->r, like it did in the normal case. I think it should be fine to just have apply set that property before calling setup_macro

laxsjo replied:
Is there anything that happens after specifically 4 arguments?

laxsjo replied:
I'm pretty sure that it isn't inlined btw

svenssonjoel replied:
> Is there anything that happens after specifically 4 arguments?

On arm I think that 3 arguments are passed via registers, but the convention switches to stack at 4. Its just a little detail but I thought, perhaps, a fun one. The application system is a very busy system, that is why I have a feeling inlining is worthwhile.

svenssonjoel replied:
> Oh, I refactored it out to this function specifically because I wanted it to not inline soo that we don't waste space. But we can maybe get rid of the macro argument, and make it come from ctx->r, like it did in the normal case. I think it should be fine to just have apply set that property before calling setup_macro

Ah I see, I will look at that again.

laxsjo replied:
I force pushed a version where it instead get's the macro form value from `ctx->r`, there was also another change you can read about. :)

laxsjo commented:
Force pushed an edited version where I changed `setup_macro` to take only 3 arguments by making what was previously read from the `macro` argument instead be read from `ctx->r`. This had no effect on the generated firmware code size.

I also factored out a common block of code in `setup_cont` and `setup_cont_sp` into `get_single_arg_or_nil`, this saved a whole 32 bytes. :) I hope that the slight performance hit to `call-cc` is worth it.

laxsjo commented:
Also, I've confirmed that my code breaks when `LBM_ALWAYS_GC` is enabled ;)

svenssonjoel commented:
> Also, I've confirmed that my code indeed does breaks when `LBM_ALWAYS_GC` is enabled ;)

Good. "apply" things are responsible for their own GCing. This is a difference between "apply" things and "fundamentals" . 
In this case it may be as easy as not popping the args off the stack until later (after the allocations that may run GC).  If you build up things incrementally, intermediate results also needs to be "protected" and this can be done either by pushing them on the stack or by, just before running gc, running lbm_mark_phase on the intermediate value. But note that if you run lbm_mark_phase, you MUST run GC you cannot do it "precautionary" just in case.. this is why macroes such as WITH_GC has an extra argument slot for things to "protect".

svenssonjoel commented:
> Force pushed an edited version where I changed `setup_macro` to take only 3 arguments by making what was previously read from the `macro` argument instead be read from `ctx->r`. This had no effect on the generated firmware code size.
> 
> I also factored out a common block of code in `setup_cont` and `setup_cont_sp` into `get_single_arg_or_nil`, this saved a whole 32 bytes. :) I hope that the slight performance hit to `call-cc` is worth it.
> 
> (In case you didn't know: you can view the changes by clicking on the "**Compare**" button next to the "laxsjo force-pushed the..." text)

Looks good. I can always try to tweak the performance after we get the functionality spot on. If you consider it fun to keep optimizing, then please do :) a couple of extra function calls if they happen every application, is probably a very bad trade for 35bytes :)  but one only knows when one benchmarks. The benchmark suite can be run again. so if you merge latest master into your branch, I can pull it and see where performance is at currently.

laxsjo commented:
> Good. "apply" things are responsible for their own GCing. This is a difference between "apply" things and "fundamentals" . In this case it may be as easy as not popping the args off the stack until later (after the allocations that may run GC). ...

Just force pushed a fix which implementents this exactly! I already had it prepared, but was just waiting for the GC tests to complete.

laxsjo commented:
> Looks good. I can always try to tweak the performance after we get the functionality spot on. If you consider it fun to keep optimizing, then please do :) a couple of extra function calls if they happen every application, is probably a very bad trade for 35bytes :) but one only knows when one benchmarks. The benchmark suite can be run again. so if you merge latest master into your branch, I can pull it and see where performance is at currently.

Just pushed another minim-improvement, this time I made `cont_application_start` call `cont_closure_application_args`, removing the need of the code handling the first iteration of the argument evaluation. It decreased the bldc firmware size by 112 bytes.

I will make sure to try running the benchmark, so that we can see how bad the performance decrease is. ;)

laxsjo commented:
Pushed a version rebased onto latest master!

svenssonjoel commented:
Bah, I am sorry, I see my silly eval_apply experiment got pushed onto master by mistake. Ill remove it now.

laxsjo commented:
> Bah, I am sorry, I see my silly eval_apply experiment got pushed onto master by mistake. Ill remove it now.

No worries, I can just rebase again! Tell me when you're ready.

svenssonjoel commented:
> > Bah, I am sorry, I see my silly eval_apply experiment got pushed onto master by mistake. Ill remove it now.
> 
> No worries, I can just rebase again! Tell me when you're ready.

Thanks, Should be ok now

laxsjo commented:
Rebased again!

svenssonjoel commented:
![image](https://github.com/user-attachments/assets/46a7b281-50f7-4fd5-aae8-01c08f1dbe64)

There is a performance hit. not  a huge one. performance has been decreasing for a while thought and it would be nice to start moving in the downwards direction again :) 

Moving the first iteration of closure_application_args into the cont, means that for all arguments there is an extra conditional executed. that is not ideal.  so its a matter here of saving some space in a place that called once, for some performance by moving that into a code that is run n times. 

One benchmark, q2, failed to run it seems.

laxsjo commented:
Just noticed that the `apply` symbol got lost from "symrepr.c" when I rebased. :P Force-pushed a fix. I assume this was why that benchmark failed.

svenssonjoel commented:
![image](https://github.com/user-attachments/assets/f1cd06ae-028d-4409-b88c-cc2c972f1268)

That missing benchmark was my fault as well :)

laxsjo commented:
> ...
> 
> There is a performance hit. not a huge one. performance has been decreasing for a while thought and it would be nice to start moving in the downwards direction again :)
> 
> Moving the first iteration of closure_application_args into the cont, means that for all arguments there is an extra conditional executed. that is not ideal. so its a matter here of saving some space in a place that called once, for some performance by moving that into a code that is run n times.
> 
> One benchmark, q2, failed to run it seems.

Interesting, if we think the cont_closure_application_args refactor isn't worth it then it's easy to remove the commit!

To put the 112 bytes into perspective, I decided to compare with how much space is left in BLDC. If i understand the memory regions/partitions table in BLDC correctly, there is currently 25236 bytes left (in flash2).
![image](https://github.com/user-attachments/assets/356a9cf8-3787-4734-9f0b-a343c43231e7)

Then the 112 bytes maybe aren't that big after all. I would be ok with removing it then. Although it would be interesting to hear what @vedderb thinks.

svenssonjoel commented:
The eval_cps file is written in a pretty odd style. It breaks all rules of proper software engineering when it come to structures programming, modularity etc. "Nice" code has been replacement with thorough testing, static analysis and benchmarking. Things may look awkward snd one may feel an unge to "cleaning it up" please think carefully before any such attempts.

laxsjo commented:
> The eval_cps file is written in a pretty odd style. It breaks all rules of proper software engineering when it come to structured programming, modularity etc. "Nice" code has been replaced with thorough testing, static analysis and benchmarking. Things may look awkward and one may feel an urge to "cleaning it up" but please think carefully before any such attempts.

Oh ok, was there anything in particular you were thinking of?

svenssonjoel commented:
> > The eval_cps file is written in a pretty odd style. It breaks all rules of proper software engineering when it come to structured programming, modularity etc. "Nice" code has been replaced with thorough testing, static analysis and benchmarking. Things may look awkward and one may feel an urge to "cleaning it up" but please think carefully before any such attempts.
> 
> Oh ok, was there anything in particular you were thinking of?

Just that the frame of mind should be to add as little as possible. If possible make it smaller but only if it maintains performance or improves it. code readability is secondary to performance and size. 

Try to minimize number of function calls and conditionals.

If possible, add the apply_apply with as little changes to anything existing as possible.  Then slowly over time we may be able to shrink code while maintaining performance.  

When it comes to code-size though, I think there are many much lower hanging fruits than trying to save bytes in eval_cps.  The various extension libraries for example probably has  a lot of potential to shrink.

laxsjo commented:
> ...
>
> Just that the frame of mind should be to add as little as possible. If possible make it smaller but only if it maintains performance or improves it. code readability is secondary to performance and size.
> 
> Try to minimize number of function calls and conditionals.
> 
> If possible, add the apply_apply with as little changes to anything existing as possible. Then slowly over time we may be able to shrink code while maintaining performance.
> 
> When it comes to code-size though, I think there are many much lower hanging fruits than trying to save bytes in eval_cps. The various extension libraries for example probably has a lot of potential to shrink.

Yeah that was definitely not the philosophy I was running with when writing this, which is completely fine! I was just focused on not bloating the code size, but your problably right that there are more low hanging fruit either way.

I can definitely put together a revised version tomorrow where I try to put these principles into practice.

svenssonjoel commented:
> > ...
> > Just that the frame of mind should be to add as little as possible. If possible make it smaller but only if it maintains performance or improves it. code readability is secondary to performance and size.
> > Try to minimize number of function calls and conditionals.
> > If possible, add the apply_apply with as little changes to anything existing as possible. Then slowly over time we may be able to shrink code while maintaining performance.
> > When it comes to code-size though, I think there are many much lower hanging fruits than trying to save bytes in eval_cps. The various extension libraries for example probably has a lot of potential to shrink.
> 
> Yeah that was definitely not the philosophy I was running with when writing this, which is completely fine! I was just focused on not bloating the code size, but your problably right that there are more low hanging fruit either way.
> 
> I can definitely put together a revised version tomorrow where I try to put these principles into practice.

That's great! Thanks a lot. 

BTW, decrease in performance is allowed when making the code more "correct" or "less buggy" and that is what we have been mostly doing for quite a while. Thus the slow decrease in performance over time that we see in the graphs. But, would be really cool to turn those trends while maintaining correctness.

laxsjo commented:
Force-pushed my revised version:
- I dropped the commit which refactored cont_closure_application_args (goodbye 9d2aa2c, you will be missed).
- I removed `get_single_arg_or_nil`, instead keeping it inlined in the source code as it was originally.
- Made `setup_cont` and `setup_cont_sp` consistent with `setup_macro` in that they expect the applied form to be stored in `ctx->r`. I'd expect this to increase performance in `cont_application_start` ever so slightly, but my main intention was to improve consistency.
- I also added an additional commit where I marked `setup_cont`, `setup_cont_sp`, and `setup_macro` as always inline. This should in theory mean that those code paths in `cont_application_start` have the same performance as they did before this PR. I thought that it would still be good to keep them as separate functions in the source code to prevent any future bugs where we accidentally only update one of the places where they are used.
  This of course significantly increased the resulting binary size by 416 bytes, but if we prioritize performance over size then this is of course a wanted change.

It would be interesting to see if there is any measurable difference in code performance between the two commits. My guess is that there wouldn't be any when running the benchmark, since it doesn't use any `macro`s or `call-cc`.

laxsjo commented:
:facepalm: Accidentally left out some changes, so I had to force-push again...

laxsjo commented on src/eval_cps.c:
```diff
@@ -4666,117 +4829,24 @@ static void cont_application_start(eval_context_t *ctx) {
-      lbm_value exp = get_cadr(get_cdr(ctx->r));
-      ctx->curr_exp = exp;
-      ctx->curr_env = expand_env;
+      pop_stack_ptr(ctx, 2);
+      setup_macro(ctx, args, env);
```
Hmm, just relalized that macro applications should still be slightly slower than before, since `cont_application_start` now pops and re-reserves a stack pointer, that could be reused in the previous implementation.

laxsjo commented:
Should also note that I'm open to make another revision where I completely remove the `setup_cont(_sp)` and `setup_macro` functions, and just copy paste the code from `cont_application_start` into `apply_apply` if you think that it's better. This would then leave `cont_application_start` untouched.

svenssonjoel commented:
Can I somehow merge in the noreturn change alone ? or can you split that out into a separate PR ? I can merge that easily and asap.

laxsjo commented:
> Can I somehow merge in the noreturn change alone ? or can you split that out into a separate PR ? I can merge that easily and asap.

Sure!

laxsjo commented:
Created #31 now. I also decided that it could include non-apply-related documentation changes.

svenssonjoel commented:
> Force-pushed my revised version:
> - I dropped the commit which refactored cont_closure_application_args (goodbye 9d2aa2c, you will be missed).
> - I removed `get_single_arg_or_nil`, instead keeping it inlined in the source code as it was originally.
> - Made `setup_cont` and `setup_cont_sp` consistent with `setup_macro` in that they expect the applied form to be stored in `ctx->r`. I'd expect this to increase performance in `cont_application_start` ever so slightly, but my main intention was to improve consistency.
> - I also added an additional commit where I marked `setup_cont`, `setup_cont_sp`, and `setup_macro` as always inline. This should in theory mean that those code paths in `cont_application_start` have the same performance as they did before this PR. I thought that it would still be good to keep them as separate functions in the source code to prevent any future bugs where we accidentally only update one of the places where they are used.
>   This of course significantly increased the resulting binary size by 416 bytes, but if we prioritize performance over size then this is of course a wanted change.
> 
> It would be interesting to see if there is any measurable difference in code performance between the two commits. My guess is that there wouldn't be any when running the benchmark, since it doesn't use any `macro`s or `call-cc`. 

This sounds nice. I'd like to benchmark it. Benjamin has additional performance tests that may exercise macros and call-cc.

vedderb commented:
One of the worse firmwares looks like this

```
Memory region         Used Size  Region Size  %age Used
           flash:       14864 B        16 KB     90.72%
          flash2:      461100 B     475120 B     97.05%
         crcinfo:           8 B          8 B    100.00%
            ram0:      119552 B       128 KB     91.21%
            ram1:          0 GB       112 KB      0.00%
            ram2:          0 GB        16 KB      0.00%
            ram3:          0 GB         0 GB      -nan%
            ram4:       62840 B        62 KB     98.98%
           libif:          2 KB         2 KB    100.00%
            ram5:          0 GB         4 KB      0.00%
            ram6:          0 GB         0 GB      -nan%
            ram7:          0 GB         0 GB      -nan%
```

So not quite as much space left. Also note that I started moving things into flash by editing the linker script and marking some of the larger functions as I ran out of flash completely the other day.

Regarding the 112 bytes vs a performance hit, I would not trade the performance hit for saving 112 bytes.

Generally, I really like having the built-in and "correct" apply, nice work on that! I have some files with weird code to work around the limitations of apply.

svenssonjoel commented:
@laxsjo do you feel like this is ready for merge?   If you align with master now I can pull branch and do a final benchcheck.

svenssonjoel commented:
I'm merging this now. Tests are happy. Thanks Rasmus, well done.

svenssonjoel commented:
Reverted the merge as it had some problems.
